### PR TITLE
Move configurations to `pyproject.toml`, use ruff

### PR DESCRIPTION
**Pull request by karosis88**
Issue: https://github.com/encode/httpcore/issues/751
Refs: https://github.com/encode/httpx/pull/2648 https://github.com/encode/httpx/pull/2680

Let's move all of our configurations to the `pyproject.toml` file so that we have a single file for our project metadata. I also added ruff as a dependency because `flake8` does not support `pyproject.toml`.

### CODE REVIEW

File: scripts/check [hunk 1]
```diff
@@ -8,8 +8,6 @@ export SOURCE_FILES="httpcore tests"
 
 set -x
 
-${PREFIX}isort --check --diff --project=httpcore $SOURCE_FILES
-${PREFIX}black --exclude '/(_sync|sync_tests)/' --check --diff --target-version=py37 $SOURCE_FILES
```
Comment (zanieb) [inline scripts/check:12]: I don't think we want to remove black here

File: scripts/check [hunk 2]
```diff
@@ -8,8 +8,6 @@ export SOURCE_FILES="httpcore tests"
...
-${PREFIX}flake8 $SOURCE_FILES
+${PREFIX}ruff check $SOURCE_FILES
```
Comment (zanieb) [inline scripts/check:11]: I think the `--show-source` option is nice here

File: scripts/lint [hunk 3]
```diff
@@ -9,8 +9,7 @@ export SOURCE_FILES="httpcore tests"
 set -x
 
 ${PREFIX}autoflake --in-place --recursive --remove-all-unused-imports $SOURCE_FILES
-${PREFIX}isort --project=httpcore $SOURCE_FILES
-${PREFIX}black --target-version=py37 --exclude '/(_sync|sync_tests)/' $SOURCE_FILES
```
[the remark this answers is not in the document]
Comment (karosis88) [reply scripts/lint:13]: Yes, we use ruff and black in httpx, but does it make sense?

[the remark this answers is not in the document]
Comment (karosis88) [reply scripts/lint:13]: Okay according to https://beta.ruff.rs/docs/faq/#is-ruff-compatible-with-black let's use both
> Yes. Ruff is compatible with [Black](https://github.com/psf/black) out-of-the-box, as long as the line-length setting is consistent between the two.
As a project, Ruff is designed to be used alongside Black and, as such, will defer implementing stylistic lint rules that are obviated by autoformatting.

**Comment by karosis88** [comment]
How did I break the `coverage` here? :-)

File: scripts/check (hunk 2 above)
Comment (karosis88) [reply scripts/check]: Yes, thank you

File: pyproject.toml [hunk 4]
```diff
@@ -66,3 +66,55 @@ path = "README.md"
...
+module = "hpack.*"
+ignore_missing_imports = true
+
+[tool.isort]
+profile = "black"
+combine_as_imports = true
+known_first_party = "httpcore,tests"
+known_third_party = "brotli,certifi,chardet,cryptography,h11,h2,hstspreload,pytest,rfc3986,setuptools,sniffio,trio,trustme,urllib3,uvicorn"
+skip = ["httpcore/_sync/,tests/_sync"]
+
```
Comment (zanieb) [inline pyproject.toml:93]: If `isort` is removed we should probably drop this section

```suggestion
```

**Review by Kludex — COMMENTED** [review]
We also removed autoflake in other encode projects, since it becomes redundant as well.

File: scripts/lint [hunk 5]
```diff
@@ -9,7 +9,7 @@ export SOURCE_FILES="httpcore tests"
 set -x
 
 ${PREFIX}autoflake --in-place --recursive --remove-all-unused-imports $SOURCE_FILES
```
Comment (zanieb) [inline scripts/lint:11]: Needs removal

Comment (karosis88) [reply scripts/lint:11]: Sure

**Review by zanieb — APPROVED** [review]

**Comment by karosis88** [comment]
@tomchristie , did we break anything? Why does this test fail while the branch is up to date?

**Comment by tomchristie** [comment]
> Why does this test fail while the branch is up to date?

Which test?

**Comment by Kludex** [comment]
> > Why does this test fail while the branch is up to date?
> 
> Which test?

It was a previous issue from click. Now it's solved.

**Comment by karosis88** [comment]
@tomchristie check [this job](https://github.com/encode/httpcore/actions/runs/5532888849/jobs/10095604921#step:7:66)

**Pull request merged**